### DISCUSSION TITLE
Crippling App Performance Issues

### DISCUSSION THREAD

**Discussion by R2bEEaton** [Budibase/budibase#8331] [Help] [answered]
Hello folks!

I set up a complex Budibase app at work over the summer. I'm a student, so I have since returned to university. I have been hearing complaints of really bad performance on the app, such as infinitely spinning data source wheels, automations not running until a few minutes later, etc. I had noticed problems like this in the past around June-August, but I was hopeful that future updates would fix it. Performance issues being faced now seem to be quite similar to what I mentioned all the way back in https://github.com/Budibase/budibase/discussions/6840.

The app is still on 1.2.52 and is becoming unusable at this point. Is there any hope for the app after updating, or will things break? It's a complex app and I haven't kept up with Budibase in the past couple of months due to school so I am unsure of how to proceed on this.

Below are graphs of performance and a short list of complaints I have been receiving:
![image](https://user-images.githubusercontent.com/34921506/196781220-f8d54575-7f07-4b10-95eb-e79425d09164.png)

- Going to the trips page, it takes a long time (>1 minute) to load all trips.
- Any screen that has to resolve the cards shows spinning circles for a while.
- Triggering the status change automation does not always update the row immediately, causing users to spam buttons and refresh.
- General navigation is slow at times.

I would prefer to keep the app on 1.2.52 to be safe and see what we can fix there but would be willing to hear any arguments for updating to fix these problems. Any advice? Thanks!

**Comment by andz-bb** [comment]
Hey @R2bEEaton, sorry to hear about the performance issues you're having! The loading times you are seeing are definitely not normal. I'm not 100% sure on the cause of your problems in this case but I would recommend you try upgrading to the latest version for sure. Anecdotally I have noticed some performance improvements on my own instance since 1.2.52 and I haven't run into any compatibility issues with my apps. Are there any parts of your app that you are particularly concerned about when thinking about updating? If there's anything specific I'd be happy to have a look!

**Reply by R2bEEaton** [reply]
Hi, thanks for your response!

I went ahead and made a backup of the server then updated Budibase. I will report on the state of the performance once regular usage resumes on Monday.

**Reply by R2bEEaton** [reply]
Hello, it's now been two weeks and the complaints are unfortunately coming back. Below is a performance graph.
![image](https://user-images.githubusercontent.com/34921506/199748816-07cb18db-a3e3-4b2c-8227-21a532d1bb44.png)

Also below is a video of the loading time: https://mediaspace.njit.edu/media/2022-11-03+10-35-15/1_su9zct0n
Funnily enough, as I set up the recording software, it seems to have sped up a little bit, but is still slow in the video.

**Reply by R2bEEaton** [reply]
@andz-bb Is there any way we could get someone to look at our app or system and find the source of these performance problems? I had a report the other day of an employee who spent 40 minutes doing a routine task in our Budibase app because of red error notifications and waiting for the loading circles to finish. The company is willing to pay for this support to get their app working again.

Otherwise we are considering moving away from Budibase and coding it ourselves.

**Reply by andz-bb** [reply]
hey @R2bEEaton, could you send the app export to support@budibase.com or on discord? I can try some investigation from my end to identify the cause of the performance issues. thanks!

**Reply by R2bEEaton** [reply]
Sent, thank you!

**Reply by andz-bb** [reply]
hey @R2bEEaton - thanks for sending that, I've had a look through the app. I think the number of data providers in some of your screens is likely causing performance issues, for example, in the trips screen which is loading several data providers in a repeater. I understand these are filtered and using Couch DB but I feel this could be done more efficiently if you're able to move these to an SQL datasource and write custom queries for them. 

Another thing that caught my eye was the automations - one of them is set on a cron trigger to run multiple queries every minute. Would it be possible to set them to run less frequently? 

I can't connect to your dbs/servers so I can't fully investigate the performance but these are a few things I spotted that could maybe be optimised. I'll be around next week so if you would like I can do some more investigation for you then.

**Reply by R2bEEaton** [reply]
Good morning, I apologize for reviving such an old thread.

First, allow me to thank you very much for having a look at our app. I see I never replied, but I assure you it was very helpful! Unfortunately, I did not have the time back then to shift the Couch DB info to a SQL data source.

This summer, I am working at the same location again for my internship, and they have asked me to revisit the app I created with Budibase. Since November 2022, would you say that there have been significant performance upgrades to the elements of our app that you believed were slow? If so, an update could be attempted. The issue is that in order for us to test it at scale involves talking with a lot of people to make sure it starts getting used again, so in lieu of that I'm basically just asking if that would be advisable without implementing any of the suggestions from your previous message OR to implement them first.

Thanks so much, and keep up the great work at Budibase!

**Reply by R2bEEaton** [reply]
Update: I have both updated Budibase AND started swapping my internal DB calls to SQL Server queries. I have noticed **huge** performance gains already. I'll keep you posted!

**Reply by R2bEEaton** [reply]
I will write a new comment explaining my findings which I will mark as the solution.

**Comment by ttelik** [comment]
I've been reading in the forum that having state variables can be a drag on performance, particularly on pages with many elements. Apparently, updating the state causes an update of every element. Perhaps that helps.

**Comment by R2bEEaton** [comment, marked as answer]
I apologize for reviving such an old discussion, but I have finally found the issue. It was solely the internal CouchDB instance. The server had many CPU cores available, and plenty of memory, but CouchDB was slowing down my app to a halt. I am not entirely sure why this was, but something like the explanations given in [this StackOverflow answer](https://stackoverflow.com/a/9815642) might be partially to blame. Regardless of that 11-year-old answer, switching from the internal CouchDB to a Microsoft SQL Server database for everything has proven all the difference.

### Before
Below is a typical week of app use from October around when employees started reporting slowdowns. Notice that the graph goes up to 3, where higher is worse.
![image](https://github.com/Budibase/budibase/assets/34921506/aa67966d-b884-4614-aac6-b9ace3655e34)

### After
Below is the app usage now over a week. Note that the graph only goes up to 1.5 now and **all** the large spikes on days 14 and 19 are from publishing app updates.
![image](https://github.com/Budibase/budibase/assets/34921506/464d2e25-6c93-49cf-8fff-e73ab7cee2ce)

### Page Load Times

|Page|Before|After
|---|---|---|
|Dashboard|850ms|800ms
|Trips|1400ms|230ms
|Assign|500ms|400ms

_Measured on the same machine within minutes of each other with Microsoft Edge DevTools. Didn't do any fancy averaging. This was done with 0 users on the Before app and very few on the After app._

It would have been so much worse if I had done this test back in October when more people were using the app, where, as reported, it would take over a minute for the Trips page to load.

### What I did
I simply moved my tables to a Microsoft SQL Server and tried to avoid using nested data providers as much as possible. I moved all of my complex data transformations to SQL queries. What was once a triple-nested data-source tree on the Trips page has been switched to just one Cards Block fed by a Custom Query.

The state variables mentioned by @ttelik didn't seem to have any effect. Hope this helps someone in the future as much as it helped me!

**Comment by xnohat** [comment]
I facing with this performance issues when my app got 10.000 users in budibase and internal table got over 5000 rows. 
After deep down in Budibase source code at the way budibase handle couchdb query, I found that all problem of budibase internal table performance is "relationship" column types. This "relationship" column type will slow down entire your budibase platform (not only individual app), reason: when budibase get a table from couchdb, it will look for any relationship column and loop through every line of relationship column and query record from "linked table". 1000 rows will query 1000 times. 
Every internal table always have 2  "relationship" columns: Created By, Updated By . These column will link to master ta_users table, when you have thousand of users, this big impact with the way budibase process these relationship column
Migrate all relationship column to text column and contain _id only, and delete all "relationship" column will dramatically increase speed all your app and your budibase self-hosted installation.

**Comment by JohannesM007** [comment]
@xnohat 
I just read this interesting message from you ;-)
How can you migrate the relationship fields into a text field and then only use the ID's. 
I would be grateful for a tip.
Tx Johannes

**Reply by R2bEEaton** [reply]
Hi @JohannesM007,

I migrated relationship fields by making use of [Primary Key / Foreign Key relationships](https://www.w3schools.com/sql/sql_foreignkey.asp). For example, I used to have a table for Drivers in Budibase that included their Name, Email, and a relationship field with all the Trips they were associated with. Now I have a Trips table that has a column called Driver, where I store the email of the driver. This could be done with another ID field if desired too. Then if I ever need the name of the driver associated with a trip, I can use a [SQL JOIN](https://www.w3schools.com/sql/sql_join.asp). All of Budibase's internal DB functionality can be achieved *and more* in SQL or with some other flavor of relational database (hence relational).

I know this was a very vague answer, but hopefully that will point you in the right direction! Perhaps @xnohat could be of more help.

**Reply by xnohat** [reply]
Yes, I have been migrate Relationship Field to Text Field with concept like @R2bEEaton mention above, deep down in the way Budibase stored relationship field in CouchDB will look like this:

`
{
  "_id": "li_ta_7a067c733c56485390c437b775e1847f_ta_users_ro_ta_7a067c733c56485390c437b775e1847f_0374530027174e068e61924e4e6d4916_ro_ta_users_us_b455c15c949f448d9c5c3f51bd7814d9_Created By_files-Created By",
  "_rev": "1-24c2f0ff41dc4dcd9ead19aa0f41c3b6",
  "type": "link",
  "doc1": {
    "tableId": "ta_7a067c733c56485390c437b775e1847f",
    "fieldName": "Created By",
    "rowId": "ro_ta_7a067c733c56485390c437b775e1847f_0374530027174e068e61924e4e6d4916"
  },
  "doc2": {
    "tableId": "ta_users",
    "fieldName": "files-Created By",
    "rowId": "ro_ta_users_us_b455c15c949f448d9c5c3f51bd7814d9"
  }
}
`

What define a relationship just a "id" key so I just using rowid ( _id in couchdb ) as data in "text field represent relationship""

**Comment by JohannesM007** [comment]
@R2bEEaton 
@xnohat 
Thank you Guys,
for the interesting information and the great tips. I will certainly dive deeper and see if I can still get performance out.
Regards, Johannes

**Comment by simon-lewis** [comment]
Similar issues for me - I have 600 records (only !) in a table with relationships to 8 other small tables (10 records each or less) (classic design pattern: metadata using IDs and lookup tables as recommended by @R2bEEaton above). With its own Relationships defined, BB opens the table in 10s or more, and a single record for editing through a sidebar in 15s to 20s. Not really usable so I will have to re-write my data sources to use JOINs. That is not so much of a problem. What will cost time - and kinda remove an advantage of using a tool like BB - is that I would have to code in each field as a lookup/dropdown in every form and table. I can skip this if I use BB as a read-only webapp but - no surprise - we need to be able to add and edit too. Pity.
